examples/cloudml-collaborative-filtering/preprocessing/preprocess.py

```diff
@@ -0,0 +1,205 @@
+    transform_fn_path: the location to save transformation outputs to.
+    user_freq: minimum frequency of a user to include it in the user vocab.
+    item_freq: minimum frequency of an item to include it in the item vocab.
+
+  Returns:
+    A pCollection of dicts, where each dict is an element of raw_data with the
+      preprocess_fn applied to it:
+      {$user_key: $user_id, $item_key: $item_id, $count_key: $count}.
+  """
+  raw_data_metadata = tft.tf_metadata.dataset_metadata.DatasetMetadata(
+      tft.tf_metadata.dataset_schema.from_feature_spec(constants.TRAIN_SPEC))
+  transformed_dataset, transform_fn = (
+      (raw_data, raw_data_metadata)
+      | beam_impl.AnalyzeAndTransformDataset(
+          lambda x: tft_fn(x, user_freq, item_freq)))
+  _ = (transform_fn
+       | "WriteTransformFn" >>
+       tft.beam.tft_beam_io.transform_fn_io.WriteTransformFn(
+           os.path.join(transform_fn_path, "transform_fn")))
+  return transformed_dataset[0]
+
+
+def _filter_data(features):
+  """Yields samples with valid users or items and then drops tft-created ids."""
+  user_id = features[constants.TFT_USER_KEY]
+  item_id = features[constants.TFT_ITEM_KEY]
+  if (user_id != constants.TFT_DEFAULT_ID
+      and item_id != constants.TFT_DEFAULT_ID):
+    yield {feature: features[feature] for feature in constants.BQ_FEATURES}
+
+
+def _clean_tags(features):
+  if (len(features[constants.TAGS_KEY]) and
+      not features[constants.TAGS_KEY][0]):
+    features[constants.TAGS_KEY] = []
+  return features
+
+
+def _split_data(examples, train_fraction=constants.TRAIN_SIZE,
+                eval_fraction=constants.EVAL_SIZE):
+  """Shuffles and splits the data into train/validation/test."""
+
+  def partition_fn(data, n_partition):  # pylint: disable=unused-argument
+    random_value = np.random.random()
+    if random_value < train_fraction:
+      return 0
+    if random_value < train_fraction + eval_fraction:
+      return 1
+    return 2
+
+  examples_split = examples | "SplitData" >> beam.Partition(partition_fn, 3)
+  return zip(["Train", "Eval", "Test"], examples_split)
+
+
+@beam.ptransform_fn
+def Shuffle(p):
+  """Shuffles the given pCollection."""
+  return (p
+          | "PairWithRandom" >> beam.Map(lambda x: (np.random.random(), x))
+          | "GroupByRandom" >> beam.GroupByKey()
+          | "DropRandom" >> beam.FlatMap(lambda (k, vs): vs))
+
+
+@beam.ptransform_fn
+def WriteOutput(p, prefix, output_dir, feature_spec, plain_text=False):
+  """Writes the given pCollection as a TF-Record.
+
+  Args:
+    p: a pCollection.
+    prefix: prefix for location tf-record will be written to.
+    output_dir: the directory or bucket to write the json data.
+    feature_spec: the feature spec of the tf-record to be written.
+    plain_text: if true, write the output as plain text instead.
+  """
+  path = os.path.join(output_dir, prefix.lower())
+  shuffled = p | "ShuffleData" >> Shuffle()  # pylint: disable=no-value-for-parameter
```

Comment:
It's a common practice to shuffle and then split data in ML, but because shuffling in Dataflow requires a `GroupByKey` and splitting was done randomly for each element, it seems more efficient to split then shuffle.

Reply:
Typically the whole dataset is shuffled and then splitting is just choosing the first 80% or so for training, the next 10% or so for validation, and the rest for test. Dataflow doesn't really lend itself to this method.

Reply:
So my understanding is typically shuffling than split is done in order to create random partitions.
Your _split_data_code is all you need (note the `partition_fn` is called for every element in pcollection which generates a random number and assigns it to a train/test/eval partition w/ probability based on your split) 

What your code currently does is create 3 random partitions and then when writing each partition in [this loop](https://github.com/GoogleCloudPlatform/professional-services/pull/266/files#diff-9fe6184ce0c744edbe5c3c40918e780aR200) you are randomly shuffling the data (within a partition ) before writing it. This is seemingly useless as the partitions were already sufficiently randomized. Is your training process (of a single partition) order dependent? 

Please feel free to ping me to discuss synchronously: jferriero@

Reply:
discussed offline, leaving shuffling to handle removing ordering from each set.